src/_types.ts

```diff
@@ -21,20 +21,15 @@ export type IterableContainer<T = unknown> = ReadonlyArray<T> | readonly [];
 
 // Inspired and largely copied from `sindresorhus/ts-extras`:
 // @see https://github.com/sindresorhus/ts-extras/blob/44f57392c5f027268330771996c4fdf9260b22d6/source/object-keys.ts
-export type ObjectKeys<T extends object> = `${Exclude<keyof T, symbol>}`;
+export type EnumeratedKeyOf<T> = `${Exclude<keyof T, symbol>}`;
 
-/**
- * Copied verbatim from `sindresorhus/ts-extras` (MIT License).
- *
- * @see https://github.com/sindresorhus/type-fest/blob/main/source/readonly-tuple.d.ts
- */
-export type ReadonlyTuple<
-  Element,
-  Length extends number,
-> = number extends Length
-  ? // Because `Length extends number` and `number extends Length`, then `Length` is not a specific finite number.
-    ReadonlyArray<Element> // It's not fixed length.
-  : BuildTupleHelper<Element, Length, []>; // Otherwise it is a fixed length tuple.
+// Object.values does only return the values paired with non-symbol keys.
+export type EnumeratedValueOf<T> = {
+  [K in keyof T]: K extends symbol
+    ? // Ignore any values that are coming from a symbol key
+      never
+    : Required<T>[K];
```

Comment:
Just curious, what is the purpose of `Required<T>[K]` here (instead of just `T[K]`)? I tried changing it to the latter and all tests still pass.

Reply:
It allows us to remove any `| undefined` from types of optional properties. The type of `T[K]` for an optional K would be `Required<T>[K] | undefined` because that is the type you'd get when you access that property in runtime (e.g. `tInstance.k`). To work around this, we need to tell typescript to treat this "access" as though the prop is not optional (by making all props in T required). If the property naturally is typed with `| undefined,` it would remain because it is not the result of optionality, this is why using `NonNullable<T[K]>` would be wrong here.

https://www.typescriptlang.org/play?exactOptionalPropertyTypes=true#code/C4TwDgpgBAglC8UDeBYAUFKBDAXFAzsAE4CWAdgOYDc6mARnoaZVAD5QCuZAJhAGbkI3GhigBjAPyNi5arSjcpBGS3ZdeAskJEBfEelCRYARgSwA2gCIslgLoiA9A8xQAehPQHw0GACYzMFZ0do7OmO6eaIY+AMwBVmIh6E4uEWheRjAALPGW3EloKeEe6VHeUABCpogAShAAjhwkREIAPDAAfFY29slhbiUZ0BX+tQ1NLdztXZbBvYX9aUOVcWONzW2dCQVFA5HRlTlrE5sz+fO7S2hAA

Notice that this acts differently when `exactOptionalPropertyTypes` is turned off in typescript. We assume the option is turned on for our typing.

Reply:
Right, but isn't the optionality being immediately reintroduced when you index by `[keyof T]` on the line below? AFAICT, `EnumeratedValueOf` would behave exactly the same if you replaced `Required<T>[K]` here with just `T[K]`. Is there an example that shows otherwise?

Reply:
`Required` here is, in fact, redundant—that's a great observation! I don't remember exactly what my thought process was when building this type. I think the intent of the `Required` was to remove `| undefined` from the types when it is the result of a prop being optional (when using `exactOptionalPropertyTypes`). That is solved in this type by removing the optionality from the keys via the mapper (`-?:`), so this is probably just an earlier attempt that I didn't remove.